arch: arm: Add code for swapping threads between secure and non-secure

oyvindronningstad opened the pull request:
This adds code to swap_helper.S which does special handling of LR when
the interrupt came from secure. The LR value is stored to memory, and
put back into LR when swapping back to the relevant thread.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

Longer explanation:

When a pendsv interrupt happens (to switch threads), LR contains a very special value (EXC_RETURN) describing the thread it came from (https://developer.arm.com/docs/100235/latest/the-cortexm33-processor/exception-model/exception-entry-and-return). The current implementation of swap_helper.S (which contains the pendsv interrupt handler) assumes a number of things about EXC_RETURN that do not hold if the CPU is executing a call into secure mode. This leads to undefined behavior if this happens, e.g. if a delayed work hits its timeout, while the CPU is in secure mode.

The current implementation leaves EXC_RETURN unchanged when switching between threads. The exception to this is when FPU is enabled. In this case, the FType bit can have different values based on whether FPU is used by the thread. This individual bit is handled correctly by swap_helper.S.

The pendsv handler normally doesn't touch the EXC_RETURN, except to set or clear a single bit to do with floating point registers. With TrustZone, there are one or two more bits that might be different between threads, the S bit (secure/nonsecure stack) and possibly the DCRS bit.

This PR introduces storing of the entire EXC_RETURN on entering the pendsv handler, and restoring the EXC_RETURN of the new thread before exiting the pendsv handler. We also set a default EXC_RETURN value when creating the thread.

This should also work for secure threads that are switched out while calling a secure function.

Why does this work?
As long as LR contains the same EXC_RETURN when a thread is switched in as when it was switched out, all the registers should be correctly popped from the relevant stack. The EXC_RETURN bits collectively specify which/how many registers are pushed, and to which stack.

How to choose the default EXC_RETURN value?
Because of the standardized way threads are initialized, there are only two valid values: 0xFFFFFFFD for secure/regular code, and 0xFFFFFFBC for non-secure code.

What are the limitations of this approach?
If one thread is executing in secure mode, no other threads can enter secure mode safely. It might work if the second thread is allowed to exit before the first thread is rescheduled, but since the NS kernel can do nothing about the S stacks, the two secure calls will be sharing stacks so the stack might overflow. As a consequence, all calls to secure mode should be prefaced with grabbing a global mutex. It might be possible to do something clever in swap_helper to not swap in a secure thread when another is running, but that is out of scope here.

oyvindronningstad committed:
arch: arm: Add code for swapping threads between secure and non-secure

This adds code to swap_helper.S which does special handling of LR when
the interrupt came from secure. The LR value is stored to memory, and
put back into LR when swapping back to the relevant thread.

Also, add special handling of FP state when switching from secure to
non-secure, since we don't know whether the original non-secure thread
(which called a secure service) was using FP registers, so we always
store them, just in case.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

ioannisg commented:
@oyvindronningstad I would appreciate if you could elaborate here, on which exact problem you are trying to address. Then we can look at the PR together.

oyvindronningstad commented:
> @oyvindronningstad I would appreciate if you could elaborate here, on which exact problem you are trying to address. Then we can look at the PR together.

I updated the PR description now. Please tell me your thoughts.

ioannisg commented:
@oyvindronningstad thanks for the longer explanation. That's interesting and inline with our earlier discussion in the downstream ticket.

I am basically going to comment on the idea, for now - not the specific implementation details.

So, as far as I understand you are trying to solve the problem of a NSPE thread properly continuing a Secure call while being context-switched out and in again, is that correct? Then, the solution you propose could theoretically work; the ARCHitecture allows you to manipulate the EXC_RETURN and return from the exception while transiting to the alternative security state. However, we must be very careful, because during the exception return, the ARCHitecture performs a series of integrity checks that need to pass, and in this case, I am not sure all these checks will always pass. For example: if the PendSV interrupt preempts a Secure interrupt then the unstacked XPSR won't match the execution mode the PE is supposed to return to. (This is of course a corner case, only valid when the Secure domain has a low priority interrupt, lower than the Non-Secure PendSV, but it is a scenario you cannot handle with this idea).

Anyway, if the integrity checks pass (**we must check this**), then your idea will work, because a NS thread's PSP (PSP_NS) remains un-touched when doing a secure call, and Zephyr Kernel ensures it is properly saved/restored. So when a thread switches back -in, to finish the secure call, when it returns from the call, the correct PSP_NS will be loaded.

>How to choose the default EXC_RETURN value?

This is indeed a "slightly" weak side of the idea, because we are constructing an EXC_RETURN value based on assumptions on how Zephyr ARCH code works for ARM. For Non-Secure PendSV we need to assume that: 
- return mode is always THREAD (never HANDLER)
- Zephyr threads use PSP (never MSP)
[the FType flag is anyways manipulated]

These are all reasonable assumptions to make.

> If one thread is executing in secure mode, no other threads can enter secure mode safely AFAIK

Yes, this is correct. We cannot read/modify the Secure stack pointer :)

> As a consequence, all calls to secure mode should be prefaced with grabbing a global mutex.

Yes, so, then, I need to question the real value of this idea. Generally, we expect NSPE to be a multi-threaded RTOS (Zephyr) so we should be locking the scheduler, anyways, before doing a secure call. Then there is no need for all this infrastructure. In summary, the idea has value only if there is a single NSPE execution context that interacts with the SPE. How probable is this scenario? Do we have such cases in our downstream project?

Side note:
The equivalent case with respect to floating-point is with FPU and FPU_SHARING:
FPU: un-shared register mode, a single thread can use the FP
FP_SHARING: many threads can use the FP

In the downstream projects I have not seen examples that enable FPU but not FP_SHARING. Because developers usually have multiple contexts that want to use the FP.

oyvindronningstad commented:
Thanks a lot for the answer.

> For Non-Secure PendSV we need to assume that:
> * return mode is always THREAD (never HANDLER)
> * Zephyr threads use PSP (never MSP)

If it's a problem, can we infer these from `CONFIG_`s and set the default EXC_RETURN accordingly?

> Generally, we expect NSPE to be a multi-threaded RTOS (Zephyr) so we should be locking the scheduler, anyways, before doing a secure call.

With this patch, we shouldn't need to lock the scheduler. If we still need to lock the scheduler for some other reason, I agree with you that it has limited utility. If we don't need to lock the scheduler (even if we do need a global mutex), time-sensitive NS threads can be executed while a long running secure service is executed (RNG for example takes ~50-100 ms), making the system more responsive.

joh06937 commented:
Sorry to randomly jump in on this, but for the last week or two I've been going down the route of trying to have Zephyr and our primary application running entirely out of Secure flash/RAM on the nRF9160 with a dedicated thread that calls into a user-written application in Non Secure flash/RAM. This PR was just pointed out to me.

The intent is that we can have our code running on our part and use TrustZone as a nice hardware protection mechanism (that requires very little intricate configuration) to allow our code to be shipped to a customer and lets them flash their own restricted application onto our device, which we would run and manage on their behalf, all without risking our IP/keys/sensitive info from being accessible in any way by them (and while providing Secure APIs for various functionality).

Unfortunately, I have run into a few roadblocks (not the least of which is the manual boundary management found in this PR and the concern over Non Secure-side interrupt management, as mentioned above). I've also had some trouble with the stacking of the floating point registers causing a fault on my part, even with a bit of extra configuration with CP11/10 in SCB->NSACR.

Anyway, my point here is that there is at least *one* person out there who's interested in a robust way to solve this issue :-)

oyvindronningstad committed:
tests: arm: Add arm_thread_swap_tz test

For testing secure->non-secure thread swapping.
This also tests that the FP context is correctly preserved
when calling a secure function.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

oyvindronningstad commented:
Rebased on top of #33510

ioannisg commented:
We want to try adding this feature for v.2.6.0, if this is accepted by the community. This may have an impact on thread scheduling on TrustZone-enabled applications.
@galak @microbuilder FYI

oyvindronningstad commented:
The failure on the an521 RPMSG sample is valid. I will debug it once I have time.

microbuilder commented on arch/arm/core/aarch32/Kconfig:
```diff
@@ -214,6 +214,31 @@ config ARM_NONSECURE_FIRMWARE
+	help
+	  When switching from a thread executing a secure service, the
+	  EXC_RETURN present in LR is specific to the secure service and the
+	  secure mode firmware in general. Ususally, the pend_sv handler doesn't
```
```suggestion
	  secure mode firmware in general. Usually, the pend_sv handler doesn't
```

oyvindronningstad commented:
I have reworked the solution now, and updated the description at the top. Please review.

oyvindronningstad commented on arch/arm/core/aarch32/Kconfig:
```diff
@@ -214,6 +214,27 @@ config ARM_NONSECURE_FIRMWARE
+	default 0xFFFFFFFD
+	help
+	  When swapping a thread out of execution, the EXC_RETURN value is
+	  stored so it ca be restored when the thread is swapped back in.
```
```suggestion
	  stored so it can be restored when the thread is swapped back in.
```

oyvindronningstad commented:
@ioannisg and I have done quite a lot of investigation into what happens if the thread calling into secure is using FP registers. The end result was the addition of code that stores all FP registers before entering the secure call if the FPU is being used by the current thread.

ioannisg commented on arch/arm/core/aarch32/swap_helper.S:
```diff
@@ -71,6 +71,10 @@ SECTION_FUNC(TEXT, z_arm_pendsv)
     ldr r2, [r1, #_kernel_offset_to_current]
 
+#ifdef CONFIG_ARM_SWAP_EXC_RETURN
+    str lr, [r2, #_thread_offset_to_exc_return]
```
I think you are not always storing the EXC_RETURN here, it might have changed due to the call to z_thread_mark_switched_out

oyvindronningstad commented:
Rebased on https://github.com/zephyrproject-rtos/zephyr/pull/34576

oyvindronningstad commented:
Rebased on master since #34576 was merged.

ioannisg committed:
arch: arm: Add config for non-blocking secure calls

Introduce a Kconfig option to allow Secure function calls to be
pre-empted.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>
Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

ioannisg commented on modules/trusted-firmware-m/interface/interface.c:
```diff
@@ -34,18 +35,61 @@ int32_t tfm_ns_interface_dispatch(veneer_fn fn,
 	k_sched_lock();
+#endif
+
+#if defined(CONFIG_FPU_SHARING) && defined(CONFIG_ARM_NONSECURE_PREEMTIBLE_SECURE_CALLS)
```
For ease of reading you can reverse the order of these two here (because above you have #ifndef CONFIG_ARM_NONSECURE_PREEMTIBLE_SECURE_CALLS)

ioannisg commented on modules/trusted-firmware-m/interface/interface.c:
```diff
@@ -34,18 +35,61 @@ int32_t tfm_ns_interface_dispatch(veneer_fn fn,
 
 	result = fn(arg0, arg1, arg2, arg3);
 
+#if defined(CONFIG_FPU_SHARING) && defined(CONFIG_ARM_NONSECURE_PREEMTIBLE_SECURE_CALLS)
```
...and same here then, so it looks uniform

ioannisg commented on tests/arch/arm/arm_thread_swap_tz/testcase.yaml:
```diff
@@ -0,0 +1,9 @@
+common:
+  filter: CONFIG_ARMV8_M_SE and CONFIG_ARM_TRUSTZONE_M and CONFIG_TRUSTED_EXECUTION_NONSECURE
```
Guess the last filter implies the first, two (in realistic use-cases)

ioannisg commented on include/arch/arm/aarch32/thread.h:
```diff
@@ -74,25 +74,46 @@ struct _thread_arch {
 	 * Bit 3: indicating whether the thread is applying the long (FLOAT)
 	 *        or the default MPU stack guard size.
+	 *
+	 * Bits 8-15: Least significant part of the EXC_RETURN value, which is
```
Proposal:
`Least significant octet of the EXC_RETURN value when a thread is switched-out. The value is loaded from LR upon the PendSV exception entry. When the thread is switched-in again, the value is restored to LR before exiting the PendSV.`

ioannisg commented on include/arch/arm/aarch32/thread.h:
```diff
@@ -74,25 +74,46 @@ struct _thread_arch {
+	union {
+		uint32_t mode;
+
+		__packed struct {
```
Do we need to do any alignment enforcement here @oyvindronningstad  
just asking.

ioannisg commented on include/arch/arm/aarch32/thread.h:
```diff
@@ -74,25 +74,46 @@ struct _thread_arch {
+		uint32_t mode;
+
+		__packed struct {
+			uint8_t bits;
```
I would keep the "mode" in the name - this is ultimately, a mode status variable.
Would use
```
mode_flags
mode_exc_return
mode_reserved
```

ioannisg commented on arch/arm/core/aarch32/Kconfig:
```diff
@@ -215,6 +215,34 @@ config ARM_NONSECURE_FIRMWARE
 
+config ARM_STORE_EXC_RETURN
+	bool
+	default y if USERSPACE || FPU_SHARING || ARCH_HAS_TRUSTED_EXECUTION
```
Why not ARM_NONSECURE_FIRMWARE @oyvindronningstad ?
ARCH_HAS does not mean anything, until users build with the feature.....

ioannisg commented on arch/arm/core/aarch32/thread.c:
```diff
@@ -101,8 +104,9 @@ void arch_new_thread(struct k_thread *thread, k_thread_stack_t *stack,
-#if defined(CONFIG_USERSPACE) || defined(CONFIG_FPU_SHARING)
+#if defined(CONFIG_ARM_STORE_EXC_RETURN)
 	thread->arch.mode = 0;
+	thread->arch.exc_return = CONFIG_ARM_DEFAULT_THREAD_EXC_RETURN;
```
I really wonder if we could use CMSIS defines  here, instead of defining the value ourselves.

ioannisg commented on arch/arm/core/aarch32/Kconfig:
```diff
@@ -215,6 +215,34 @@ config ARM_NONSECURE_FIRMWARE
+	  This is needed when switching between threads that differ in either
+	  FPU usage or security domain.
+
+config ARM_DEFAULT_THREAD_EXC_RETURN
```
Why does it need to be a Kconfig?
Could we just define the values in arch.h for Cortex-M?

ioannisg replied:
if we decide to keep it here: could we add a depends on ARM_STORE_EXC_RETURN?

ioannisg commented on arch/arm/core/aarch32/thread.c:
```diff
@@ -23,6 +23,9 @@
 #endif
 
+/* bit [4] allocate stack for floating-point context: 0=done 1=skipped  */
+#define EXC_RETURN_FTYPE           (0x00000010UL)
```
might be able to take this from CMSIS?

ioannisg commented on arch/arm/core/aarch32/swap_helper.S:
```diff
@@ -71,6 +71,10 @@ SECTION_FUNC(TEXT, z_arm_pendsv)
     ldr r1, =_kernel
     ldr r2, [r1, #_kernel_offset_to_current]
 
+#if defined(CONFIG_ARM_STORE_EXC_RETURN)
```
Please, add a comment:
E.g. `Store LSB of LR (EXC_RETURN) to the thread's mode octet reserved for this.`

ioannisg replied:
looks fine

ioannisg commented on tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:
```diff
@@ -25,6 +25,9 @@
 #define SWAP_RETVAL        0x1234
 
+/* bit [4] allocate stack for floating-point context: 0=done 1=skipped  */
+#define EXC_RETURN_FTYPE           (0x00000010UL)
```
Again, see if there are any CMSIS macros related (I think there are)

ioannisg commented on tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:
```diff
@@ -256,7 +259,7 @@ static void alt_thread_entry(void)
 	 */
-	zassert_true((_current->arch.mode & CONTROL_FPCA_Msk) == 0,
+	zassert_true((_current->arch.exc_return & EXC_RETURN_FTYPE) != 0,
 		"Alt thread FPCA flag not clear at initialization\n");
```
please, adapt also the ASSERT comment.

ioannisg commented on tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:
```diff
@@ -276,7 +279,7 @@ static void alt_thread_entry(void)
 
-	zassert_true((p_ztest_thread->arch.mode & CONTROL_FPCA_Msk) != 0,
+	zassert_true((p_ztest_thread->arch.exc_return & EXC_RETURN_FTYPE) == 0,
 		"ztest thread mode FPCA flag not updated at swap-out: 0x%0x\n",
```
Same here

ioannisg commented on tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:
```diff
@@ -447,7 +450,7 @@ void test_arm_thread_swap(void)
 #if defined(CONFIG_FPU) && defined(CONFIG_FPU_SHARING)
 	/* The main test thread is not (yet) actively using the FP registers */
-	zassert_true((_current->arch.mode & CONTROL_FPCA_Msk) == 0,
+	zassert_true((_current->arch.exc_return & EXC_RETURN_FTYPE) != 0,
```
and here

ioannisg commented on tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:
```diff
@@ -476,7 +479,7 @@ void test_arm_thread_swap(void)
 	 * flag is not updated until the next context switch.
 	 */
-	zassert_true((_current->arch.mode & CONTROL_FPCA_Msk) == 0,
+	zassert_true((_current->arch.exc_return & EXC_RETURN_FTYPE) != 0,
```
and here

ioannisg commented on tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c:
```diff
@@ -700,7 +703,7 @@ void test_arm_thread_swap(void)
 	 * flag and MPU GUARD flag are now updated.
 	 */
-	zassert_true((_current->arch.mode & CONTROL_FPCA_Msk) != 0,
+	zassert_true((_current->arch.exc_return & EXC_RETURN_FTYPE) == 0,
```
and here

ioannisg commented on tests/arch/arm/arm_thread_swap_tz/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+
+cmake_minimum_required(VERSION 3.13.1)
+
+set(QEMU_KERNEL_OPTION "-device;loader,file=${CMAKE_BINARY_DIR}/tfm_merged.hex")
```
needed?

oyvindronningstad reviewed (commented):
Thanks @ioannisg, all comments should be addressed.

ioannisg reviewed (approved):
Looks good to me now, @oyvindronningstad , thanks for this work. If issues pop up we could sort them out during the stabilization period.

ioannisg commented on arch/arm/core/aarch32/swap_helper.S:
```diff
@@ -201,6 +197,11 @@ out_fp_endif:
+
+#if defined(CONFIG_ARM_STORE_EXC_RETURN)
+    /* Restore EXC_RETURN value. */
+    ldrsb lr, [r2, #_thread_offset_to_mode_exc_return]
```
@oyvindronningstad just a question - why this is LDRSB, and not LDRB?

oyvindronningstad replied:
LDRSB does sign extension which gives the initial 0xFFFFFF.

ioannisg commented on modules/trusted-firmware-m/interface/interface.c:
```diff
@@ -34,18 +35,61 @@ int32_t tfm_ns_interface_dispatch(veneer_fn fn,
+		context_saved = true;
+
+		/* Disable FPCA so no stacking of FP registers happens in TFM. */
+		__set_CONTROL(CONTROL & ~CONTROL_FPCA_Msk);
```
Do we need an ISB here?

oyvindronningstad replied:
Not sure, I can look into it.

carlocaione commented on arch/arm/core/offsets/offsets_aarch32.c:
```diff
@@ -32,12 +32,15 @@
 GEN_OFFSET_SYM(_thread_arch_t, swap_return_value);
 
-#if defined(CONFIG_USERSPACE) || defined(CONFIG_FPU_SHARING)
+#if defined(CONFIG_ARM_STORE_EXC_RETURN) || defined(CONFIG_USERSPACE)
 GEN_OFFSET_SYM(_thread_arch_t, mode);
+#endif
```
are you using `_thread_offset_to_mode` when `CONFIG_ARM_STORE_EXC_RETURN` is defined?

ioannisg replied:
I think only when CONFIG_FPU_SHARING is defined, right @oyvindronningstad ? So, when we only care about S-NS transitions, without USERSPACE and without FPU_SHARING, we do not use the mode offset, only the mode_exc_return..

oyvindronningstad replied:
Yep, I think that's right. Changed to 
```suggestion
#if defined(CONFIG_FPU_SHARING) || defined(CONFIG_USERSPACE)
GEN_OFFSET_SYM(_thread_arch_t, mode);
#endif
```

ioannisg replied:
thanks
@carlocaione ^^

ioannisg committed:
doc: release notes: release notes entry for non-blocking secure calls

Add an entry in the release notes regarding the
support for non-blocking secure calls in Zephyr.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented on arch/arm/core/aarch32/Kconfig:
```diff
@@ -215,6 +215,24 @@ config ARM_NONSECURE_FIRMWARE
 
+config ARM_STORE_EXC_RETURN
+	bool
+	default y if FPU_SHARING || ARM_NONSECURE_FIRMWARE
```
@oyvindronningstad do we need to change the `ARM_NONSECURE_FIRMWARE `to `ARM_NONSECURE_PREEMTIBLE_SECURE_CALLS` in the dependencies here?

In other words, when we do not support preemptible calls, and we do not have FPU, do we need to store the EXC_RETURN?

oyvindronningstad replied:
Yes, we need it to correctly keep the S/ES/DCRS flags.

ioannisg replied:
I am not sure we need to do anything if we do not switch security domains at swap. If we only switch from NS threads to NS threads, I mean, why do we need to store and restore the EXC_RETURN?

ioannisg commented on arch/arm/core/aarch32/Kconfig:
```diff
@@ -215,6 +215,24 @@ config ARM_NONSECURE_FIRMWARE
+	  This is needed when switching between threads that differ in either
+	  FPU usage or security domain.
+
+config ARM_NONSECURE_PREEMTIBLE_SECURE_CALLS
```
@oyvindronningstad by they way I just realized there is a typo here, sorry.
Should  be ARM_NONSECURE_PREEMPTIBLE_SECURE_CALLS

oyvindronningstad committed:
trusted-firmware-m: Store FP context before entering secure calls

When ARM_NONSECURE_PREEMPTIBLE_SECURE_CALLS is enabled, if FPU is
being used (CONTROL.FPCA == 1), store all FP registers before
entering the secure function, and restore them afterwards.

This is needed if any NS thread or ISR that interrupts the secure
function uses FP registers. If they do, a secure UsageFault occurs
unless this change is applied.

This allows k_sched_lock() and k_sched_unlock() to be dropped when
ARM_NONSECURE_PREEMPTIBLE_SECURE_CALLS is enabled.

Enable ARM_NONSECURE_PREEMPTIBLE_SECURE_CALLS by default when
building TF-M.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

microbuilder reviewed (approved):
Nice work on this @oyvindronningstad

ioannisg commented:
I'll add a patch to fix the test failure for the STM boards - apparently the overlay is not correct. To make this easy to track, I will not force-push, so just apply the patch on top of the PR. Please don't merge.

oyvindronningstad commented:
> I'll add a patch to fix the test failure for the STM boards - apparently the overlay is not correct. To make this easy to track, I will not force-push, so just apply the patch on top of the PR. Please don't merge.

Which test failure? Looks like the CI is green.

ioannisg commented:
> > I'll add a patch to fix the test failure for the STM boards - apparently the overlay is not correct. To make this easy to track, I will not force-push, so just apply the patch on top of the PR. Please don't merge.
> 
> Which test failure? Looks like the CI is green.

The overlay is not correct, and the NS test does not boot. After changing the overlay the test is able to run, however it fails. Attaching log

```
*** Booting Zephyr OS build zephyr-v2.5.0-3129-gb84e7d4cdcb2  ***
Running test suite test_thread_swap
===================================================================
START - test_thread_swap
    Assertion failed at ../s
    Assertion failed at ../src/main.c:41: work_func: (main_thread->arch.mode_exc_return & EXC_RETURN_S is false)
EXC_RETURN not secure: 0xac
 FAIL - test_thread_swap in 0.19 seconds
===================================================================
Test suite test_thread_swap failed.
===================================================================
PROJECT EXECUTION FAILED
```

oyvindronningstad commented:
>  After changing the overlay the test is able to run, however it fails. 

That error means there's probably some timing issue in the test. The work function didn't manage to interrupt the secure service. Likely the crypto is faster and finishes before the work fires.

ioannisg commented:
> > After changing the overlay the test is able to run, however it fails.
> 
> That error means there's probably some timing issue in the test. The work function didn't manage to interrupt the secure service. Likely the crypto is faster and finishes before the work fires.

I understand that. What is your proposal to fix this? Or should we remove the ST platforms from the allow-list?

oyvindronningstad commented:
Try decreasing the delay on the work

ioannisg committed:
tests: arm_thread_swap_tz: decrease work delay to ensure test passing

Decrease the submitted work delay, to ensure that the PSA
crypto operation will be preempted when the work fires. The
modification is required for devices with fast crypto
operations. Also minor corrections to the test ase name,
so it is not the same with the other arm_thread_swap test.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
tests: arm_thread_swap_tz: correct dt overlays for stm32 boards

Correct thee dt overlays for the STM32 boards, so the Zephyr
image starting address is in accordance with what TF-m expects
it to be.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
Test is passing now on
- nrf5340
- nrf9160
- nucleo_l552ze_q
- stm32l562e_q
- mps2_an521

microbuilder commented:
A bit late to the party, but tests also passing on `lpcxpresso55s69_ns`

MZachmann commented:
on Windows I'm getting a warning with gnu arm (q4 2020) that __packed in the struct near line 114 is being ignored and it needs to be after the struct keyword. It's fine with struct __packed.

ioannisg commented:
> etting a warning with gnu arm (q4 2020) that __packed in the stru

@oyvindronningstad you might want to take a look?

oyvindronningstad commented:
> 
> 
> on Windows I'm getting a warning with gnu arm (q4 2020) that __packed in the struct near line 114 is being ignored and it needs to be after the struct keyword. It's fine with struct __packed.

@MZachmann Thanks, do you mind checking that this removes your warning? https://github.com/zephyrproject-rtos/zephyr/pull/35031
